File: Algorithms/32. Longest Valid Parentheses/python/solution.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        maxLength, left, right = 0, 0, 0

        for value in s:
            if value == "(":
                left += 1
            else:
                right += 1

            if left == right:
                maxLength = max(maxLength, left + right)
            elif right > left:
                left, right = 0, 0

        left, right = 0, 0
        for i in range(len(s) - 1, -1, -1):
            if s[i] == "(":
                left += 1
            else:
                right += 1

            if left == right:
                maxLength = max(maxLength, left + right)
            elif left > right:
                left, right = 0, 0

        return maxLength

File: Algorithms/32. Longest Valid Parentheses/golang/solution.go
func longestValidParentheses(s string) int {
	max := 0
	left := 0
	right := 0

	for i := 0; i < len(s); i++ {
		if s[i] == '(' {
			left++
		} else {
			right++
		}

		if left == right {
			if left+right > max {
				max = right + left
			}
		} else if right > left {
			right = 0
			left = 0
		}
	}

	left = 0
	right = 0
	for i := len(s) - 1; i >= 0; i-- {
		if s[i] == '(' {
			left++
		} else {
			right++
		}

		if left == right {
			if left+right > max {
				max = right + left
			}
		} else if left > right {
			right = 0
			left = 0
		}
	}
	return max
}
